cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java

```diff
@@ -44,18 +49,30 @@ public boolean equals(Object o) {
 
     SourceControlMeta that = (SourceControlMeta) o;
 
-    return Objects.equals(fileHash, that.fileHash);
+    return Objects.equals(fileHash, that.fileHash)
+        && Objects.equals(commitId, that.commitId)
+        && Objects.equals(lastSyncedAt, that.lastSyncedAt);
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(fileHash);
+    return Objects.hash(fileHash, lastSyncedAt, commitId);
   }
 
   @Override
   public String toString() {
     return "SourceControlMeta{"
         + "fileHash='" + fileHash + '\''
+        + "commitId='" + commitId + '\''
+        + "lastSyncedAt='" + lastSyncedAt + '\''
         + '}';
   }
+
+  public String getCommitId() {
```

Comment:
Are we handling existing data that was serialized without these fields? These would need to be nullable right?

Reply:
GSON deserilaizes it to null if missing as both of the fields are objects, I have verified the same. I opted not to add Nullable as it would signify in future usage that this fields can be null but it should never be initialised with null going forward.

Reply:
I had originally opted not to add nullable to constructor as GSON would anyway deserialize as Null ( tested) and for future usage no one should call the constructor with null.
But the getters should have Nullable annotation. I have added the same

Reply:
I don't see the annotation, were all your changes pushed?

Reply:
also looks like AppLifecycleService.updateSourceControlMeta() calls the constructor, so it can potentially be null in constructor too.